### PR TITLE
Webhook v1.5: PlutoSlidersServer with self-updating Capabilities (3/2)

### PR DESCRIPTION
**Pull request by pankgeorg**
(Same as Webhook v2, but without unnesessary functions that duplicate functionality)

This PR introduces the following changes:

1. src/Actions.jl:L57: Add FileWatching capabilities: Watch every running session
2. src/Actions.jl:L99: watchnotebookready: block until notebook is ready. Not used now, see comment@L153
3. src/Actions.jl:L132: renew_session!: What do do when file changes. Invoked by L61 (@asynclog watch) and the webhook.
4. src/Actions.jl:L168: remove_from_session!: How to stop a notebook from running
5. src/PlutoSliderServer.jl:L120: register webhook (see webhook)
6. src/Types.jl:L5: move get_configuration code from PlutoSliderServer to this file (so we can import to Webhook and restart if config changed)
7. src/UpdateFromFile.jl: Copy code from Pluto.jl#watch-notebook-file-1 so PlutoSliderServer can use it without pluto implementing (also see #2)
8. src/Webhook.jl: On POST @ /github_webhook (see below)

# How the webhook works

NOTE: The webhook assumes you already are running the julia process in a folder which is a git repository. (.git exists)

1. Listen for POSTS @ /github_webhook 
2. If you have provided a secret (HIGHLY RECOMMENDED) it will be validated. If validation fails, 501.
3. git pull runs
4. (PlutoSliderServer picks up changes in changed files and renews sessions)
5. webhook
	a. starts new files
	b. stops old files
	c. renews changed files (if picked up by filesystem, one of them will return doing nothing)

### CODE REVIEW

File: src/Actions.jl [hunk 1]
```diff
@@ -67,7 +92,106 @@ module Actions
...
+            read(joinpath(settings.SliderServer.start_dir, path), String)
+        catch e
+            @warn "notebook deleted; removing"
+            remove_from_session!(notebook_sessions, server_session, notebook_sessions[i].hash)
+            return
+        end
+        new_hash = path_hash(path)
+        session = notebook_sessions[i]
+        if new_hash == session.hash
+           println("Renewing unnecessary; returning!")
```
Comment (pankgeorg) [inline src/Actions.jl:149]: return!!!

File: src/UpdateFromFile.jl [hunk 2]
```diff
@@ -0,0 +1,50 @@
+"""
```
Comment (pankgeorg) [inline src/UpdateFromFile.jl:1]: Import this from Pluto

File: src/Webhook.jl [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+            if get(ENV, "GITHUB_SECRET", "") !== ""
+                security_test = validate_github_headers(request, ENV["GITHUB_SECRET"])
+                if !security_test
+                    return HTTP.Response(501, "Not authorized!")
+                end
+            end
+
+            # github_url = get(get(JSON.parse(String(request.body)), "repository", Dict()), "html_url", nothing)
+            @async try
+                run(`git pull`)
```
Comment (pankgeorg) [inline src/Webhook.jl:41]: `git reset --hard` (because notebooks may change other notebooks? just be on the safe side)

File: src/Webhook.jl [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+                # run(`git checkout`)
+                config_toml_path = joinpath(Base.active_project() |> dirname, "PlutoDeployment.toml")
+                new_settings = get_configuration(config_toml_path)
+                @info new_settings
+                @info new_settings == settings
+                # TODO: Restart if settings changed
+                old_paths = map(notebook_sessions) do sesh
+                    sesh isa RunningNotebookSession ? sesh.path : nothing
```
Comment (pankgeorg) [inline src/Webhook.jl:49]: Handle all the NotebookSession types:
`QueuedNotebookSessions` and `FinishedNotebookSession`

File: src/Webhook.jl [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+                end
+
+                for path in to_renew
+                    session, jl_contents, original_state = renew_session!(notebook_sessions, server_session, path, settings)
+                    if path ∉ settings.Export.exclude
+                        generate_static_export(path, settings, original_state, settings.Export.output_dir, jl_contents)
+                    end
+                end
+
+                for path in to_start
```
Comment (pankgeorg) [inline src/Webhook.jl:81]: Add Queued Sessions somewhere

**Comment by pankgeorg** [comment]
Ideas: 
1. lock list of sessions
2. create a `desired` list of notebooks/state and `while(not equal) make equal else sleep`

File: src/Actions.jl [hunk 6]
```diff
@@ -39,6 +54,16 @@ module Actions
                     Pluto.SessionActions.shutdown(server_session, notebook)
                 end
                 if keep_running
+                    @asynclog while true
```
Comment (pankgeorg) [inline src/Actions.jl:57]: Let's comment this out!

File: src/Webhook.jl [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+        the server will try to
+            - reload all changed files,
+            - stop all deleted files and
+            - start any new files
+        respecting the settings (exlusions etc.)
+
+        TODO: restart julia process if settings (assumed to be at
+        `pluto-deployment-environment/PlutoDeployment.toml`) change.
+        """
+        function pull(request::HTTP.Request)
```
Comment (pankgeorg) [inline src/Webhook.jl:30]: 3 parts: 
- authenticate
- pull
- enter update delete (factor this out?)

File: src/Webhook.jl [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+                # run(`git checkout`)
+                config_toml_path = joinpath(Base.active_project() |> dirname, "PlutoDeployment.toml")
+                new_settings = get_configuration(config_toml_path)
+                @info new_settings
+                @info new_settings == settings
+                # TODO: Restart if settings changed
+                old_paths = map(notebook_sessions) do sesh
+                    sesh isa RunningNotebookSession ? sesh.path : nothing
+                end
+                old_hashes = map(notebook_sessions) do sesh
```
Comment (pankgeorg) [inline src/Webhook.jl:51]: filter / map / do a comprehension

File: src/Types.jl [hunk 9]
```diff
@@ -27,16 +28,64 @@ module Types
         original_state
     end
     
-    
+    Base.@kwdef struct NotebookSessionList
```
Comment (pankgeorg) [inline src/Types.jl:31]: @fonsp take a look here 😇

**Comment by fonsp** [comment]
TODO:
- [ ] Show progress in logs (e.g. `[12/29] ready`)
- [ ] `git reset`? Don't delete html files? Require separate output folder?
- [ ] Check for concurrency stuff

**Comment by fonsp** [comment]
I made some changes:
- Git sync is now based on watching the `start_dir` for changes recursively, using [BetterFileWatching.jl](https://github.com/JuliaPluto/BetterFileWatching.jl). This means that we can test the sync feature without git, and git is decoupled. You could also use Google Drive as your deployment platform :)
- There is a function `refresh_until_synced` that will rescan the folder for notebook files, and then `process` **one** notebook that needs processing.
- more

**Comment by fonsp** [comment]
I renamed the branch which apparently means that we have to make a new PR! #31

**Pull request closed without merging**